src/cprover/chc_db.h

```diff
@@ -0,0 +1,233 @@
+//
+// Created by Yakir Vizel on 5/27/24.
+//
+
+#ifndef CBMC_CHC_DB_H
+#define CBMC_CHC_DB_H
+
+#include <util/mathematical_expr.h>
+#include <util/std_expr.h>
+#include <util/find_symbols.h>
+
+#include <vector>
+#include <map>
+#include <set>
+#include <functional>
+
+class chc_dbt;
+
+/*
+ * A horn clause.
+ * This class is simply a wrapper around a forall_exprt with a few utilities:
+ * 1. Getting the body of a clause
+ * 2. Getting the head of a clause
+ * 3. Checking if a clause is a fact or a query
+ * 4. Getting used relations (the predicates) or function applications
+ *    (their instantiations) in a clause.
+ */
+class horn_clauset
+{
+  forall_exprt m_chc;
+
+public:
+  horn_clauset(forall_exprt f) : m_chc(f) {}
+
+  horn_clauset(std::vector<symbol_exprt> & vars, exprt clause) : m_chc(vars, clause) {
+
+  }
+
+  const forall_exprt & get_chc() const
+  {
+    return m_chc;
+  }
+
+  const exprt* body() const {
+    if (can_cast_expr<implies_exprt>(m_chc.where()))
+    {
+      return &to_implies_expr(m_chc.where()).op0();
+    }
+    return &m_chc.where();
+  }
+
+  const exprt* head() const {
+    if (can_cast_expr<implies_exprt>(m_chc.where()))
+    {
+      return &to_implies_expr(m_chc.where()).op1();
+    }
+    return nullptr;
+  }
+
+  bool is_fact() const {
+    auto b = body();
+    bool not_fact = false;
+    b->visit_pre(
+      [&not_fact](const exprt &expr)  {
+        if(can_cast_expr<function_application_exprt>(expr))
+        {
+          not_fact = true;
+        }
+      });
```

Comment:
I think you'd be better off using `depth_iteratort` so that you can abort the traversal as soon as a you hit a `function_application_exprt`.